contracts/contracts/harvest/BaseHarvester.sol

```diff
@@ -104,82 +137,169 @@ abstract contract BaseHarvester is Governable {
         if (
-            /* oldUniswapAddress == address(0) when there is no pre-existing
+            /* oldRouterAddress == address(0) when there is no pre-existing
              * configuration for said rewards token
              */
-            oldUniswapAddress != address(0) &&
-            oldUniswapAddress != _uniswapV2CompatibleAddr
+            oldRouterAddress != address(0) &&
+            oldRouterAddress != newRouterAddress
         ) {
-            token.safeApprove(oldUniswapAddress, 0);
+            token.safeApprove(oldRouterAddress, 0);
         }
 
-        // Give Uniswap infinite approval when needed
-        if (oldUniswapAddress != _uniswapV2CompatibleAddr) {
-            token.safeApprove(_uniswapV2CompatibleAddr, 0);
-            token.safeApprove(_uniswapV2CompatibleAddr, type(uint256).max);
+        // Give SwapRouter infinite approval when needed
+        if (oldRouterAddress != newRouterAddress) {
+            token.safeApprove(newRouterAddress, 0);
+            token.safeApprove(newRouterAddress, type(uint256).max);
+        }
+
+        SwapPlatform _platform = tokenConfig.platform;
+        if (_platform == SwapPlatform.UniswapV2Compatible) {
+            uniswapV2Path[_tokenAddress] = _decodeUniswapV2Path(
+                swapData,
+                _tokenAddress
+            );
+        } else if (_platform == SwapPlatform.UniswapV3) {
+            uniswapV3Path[_tokenAddress] = _decodeUniswapV3Path(
+                swapData,
+                _tokenAddress
+            );
+        } else if (_platform == SwapPlatform.Balancer) {
+            balancerPoolId[_tokenAddress] = _decodeBalancerPoolId(swapData);
+        } else if (_platform == SwapPlatform.Curve) {
+            curvePoolData[_tokenAddress] = _decodeCurvePool(
+                swapData,
+                newRouterAddress,
+                _tokenAddress
+            );
+        } else {
+            revert("Unknown SwapPlatfrom");
         }
 
         emit RewardTokenConfigUpdated(
             _tokenAddress,
-            _allowedSlippageBps,
-            _harvestRewardBps,
-            _uniswapV2CompatibleAddr,
-            _liquidationLimit,
-            _doSwapRewardToken
+            tokenConfig.allowedSlippageBps,
+            tokenConfig.harvestRewardBps,
+            _platform,
+            newRouterAddress,
+            swapData,
+            tokenConfig.liquidationLimit,
+            tokenConfig.doSwapRewardToken
+        );
+    }
+
+    function _decodeUniswapV2Path(bytes calldata data, address token)
+        internal
+        view
+        returns (address[] memory path)
+    {
+        (path) = abi.decode(data, (address[]));
+
+        // Do some validation
+        require(
+            path.length >= 2 &&
+                path[0] == token &&
+                path[path.length - 1] == baseTokenAddress,
+            "Invalid Uniswap V2 path"
+        );
+    }
+
+    function _decodeUniswapV3Path(bytes calldata data, address token)
```

Comment:
I think you might be able to uses slices on this calldata data + abi.decode to make this be a bit cleaner and simpler.

Off the top of my head: `abi.decode(data[0:20], (address))` or something.

Reply:
Apparently you couldn't use `abi.decode` with sliced bytes (it's treated as type `bytes memory slice`). So I had to do explicit conversions after slicing them. Should be much shorter and cleaner now, thanks for the tip